utilities/backupable/backupable_db.cc

```diff
@@ -1733,25 +1733,51 @@ Status BackupEngineImpl::BackupMeta::StoreToFile(bool sync) {
   if (!app_metadata_.empty()) {
     std::string hex_encoded_metadata =
         Slice(app_metadata_).ToString(/* hex */ true);
+
+    // +1 to accomodate newline character
+    size_t hex_meta_strlen = kMetaDataPrefix.ToString().length() + hex_encoded_metadata.length() + 1;
+    if (hex_meta_strlen >= buf_size) {
+      return Status::Corruption("Buffer too small to fit backup metadata");
+    }
+    else if (len + hex_meta_strlen >= buf_size) {
+      backup_meta_file->Append(Slice(buf.get(), len));
+      buf.release();
+      unique_ptr<char[]> new_reset_buf(new char[max_backup_meta_file_size_]);
+      buf.swap(new_reset_buf);
+      len = 0;
+    }
     len += snprintf(buf.get() + len, buf_size - len, "%s%s\n",
                     kMetaDataPrefix.ToString().c_str(),
                     hex_encoded_metadata.c_str());
-    if (len >= buf_size) {
-      return Status::Corruption("Buffer too small to fit backup metadata");
-    }
   }
-  len += snprintf(buf.get() + len, buf_size - len, "%" ROCKSDB_PRIszt "\n",
-                  files_.size());
-  if (len >= buf_size) {
-    return Status::Corruption("Buffer too small to fit backup metadata");
+
+  char writelen_temp[19];
+  if (len + sprintf(writelen_temp, "%" ROCKSDB_PRIszt "\n", files_.size()) >= buf_size) {
+    backup_meta_file->Append(Slice(buf.get(), len));
+    buf.release();
```

Comment:
The `release()` causes ownership of the underlying array to be given up by the unique_ptr so that memory will not be freed. You can see this by building with ASAN and doing a backup of a DB with many files. One trick I used was setting `max_backup_meta_file_size_ = 10240` to make it easier to hit the limit.

```
$ COMPILE_WITH_ASAN=1 make -j64 db_bench ldb sst_dump
$ TEST_TMPDIR=/dev/shm ./db_bench -benchmarks=fillseq -write_buffer_size=4096 -max_background_jobs=24 -compression_type=none -memtablerep=vector -allow_concurrent_memtable_write=false
$ ./ldb backup --db=/dev/shm/dbbench/ --backup_dir=./backup-tmp
open db OK
open backup engine OK
create new backup OK


=================================================================
==4128042==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 61440 byte(s) in 6 object(s) allocated from:
    #0 0x7f6c1c37b09a in operator new[](unsigned long) (/usr/local/fbcode/gcc-5-glibc-2.23/lib/libasan.so.2+0x9b09a)
    #1 0xa94461 in rocksdb::BackupEngineImpl::BackupMeta::StoreToFile(bool) utilities/backupable/backupable_db.cc:1775
    #2 0xaa7a91 in rocksdb::BackupEngineImpl::CreateNewBackupWithMetadata(rocksdb::DB*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, std::function<void ()>) utilities/backupable/backupable
_db.cc:878
    #3 0xab0ae4 in rocksdb::BackupEngine::CreateNewBackup(rocksdb::DB*, bool, std::function<void ()>) include/rocksdb/utilities/backupable_db.h:271
    #4 0xce4378 in rocksdb::BackupCommand::DoCommand() tools/ldb_cmd.cc:2758
    #5 0xcd2ae9 in rocksdb::LDBCommand::Run() tools/ldb_cmd.cc:267
    #6 0xd03d4e in rocksdb::LDBCommandRunner::RunCommand(int, char**, rocksdb::Options, rocksdb::LDBOptions const&, std::vector<rocksdb::ColumnFamilyDescriptor, std::allocator<rocksdb::ColumnFamilyDescriptor> > const*) tools/ldb_tool.cc:
115
    #7 0xd062c1 in rocksdb::LDBTool::Run(int, char**, rocksdb::Options, rocksdb::LDBOptions const&, std::vector<rocksdb::ColumnFamilyDescriptor, std::allocator<rocksdb::ColumnFamilyDescriptor> > const*) tools/ldb_tool.cc:127
    #8 0x40e0df in main tools/ldb.cc:12
    #9 0x7f6c1b059857 in __libc_start_main (/usr/local/fbcode/gcc-5-glibc-2.23/lib/libc.so.6+0x20857)
    #10 0x51c0cf  (/data/users/andrewkr/rocksdb/ldb+0x51c0cf)
```

Reply:
Thanks for the feedback. I had earlier planned on using buf.reset() to empty the buffer. I've realised that release() just yields ownership of pointer.